### ISSUE TITLE
Switch to storage by scope instead of resource server

### ISSUE THREAD

**Issue by NickolausDS** [fair-research/native-login#27]
_This is an exploration into advantages/disadvantages of this feature, we may not actually want to implement this_

Currently we store tokens by resource server, which is simple and works for simple use cases. 

```
{
  "auth.globus.org": {
    "scope": "profile openid email",
    ...
  }
}
```

More complex use cases may prefer to store tokens by scope, and possibly even store multiple tokens for the same resource server under different scope names. A solution may look like this:

```
{
  "profile": {
    "access_token": "<access_token>",
    "resource_server: "auth.globus.org",
    ...
  },
  "openid": {
    "access_token": "<access_token>",
    "resource_server: "auth.globus.org",
    ...
  }
}
```

Advantages

* Would allow for more complex use cases, grab separate access tokens for different scopes

Disadvantages

* More complex storage
* ~Scope names are not unique, and we'd need to devise a better storage mechanism to account for that. For example, Transfer and Search both have the `all` scope, so we may need to store those by resource server anyway in order not to accidentally overwrite them.~ Scope names should always be unique.  
    * ~This poses a challenge for supporting ConfigParser storage, due to lack of ability to nest config levels. We could just optionally require ConfigObj instead.~ This is still a valid concern, but might be addressed by packing more info into the "key" for ConfigParser storage. An implementation using ConfigObj nested storage would still be nice.
* Fetching tokens by resource server will be ambiguous, and may result in a list of possible tokens.

### Alternatives

Requesting additional scopes for the same resource server could cause Native Login to combine those into a single token request. For example:

```
client.login(requested_scopes=['profile'])
client.login(requested_scopes=['openid'])
```
Could cause the second call to `login()` to additionally request a scope to profile and save the resulting token. 

Advantages

* results in simpler usage, storage, and more flexible tokens. 

Disadvantages

* In the example above, the first login call to fetch the `profile` scope may still be in use. From the native login packages's perspective, this token may not be needed or could still be in use. After the second login call to get `['profile', 'openid']`, the first `profile` token will likely still be active but `orphaned` after the tokens are saved over. There are a few ways to mitigate this:
     * Rely on the user to revoke old tokens before requesting new ones
     * Add a new storage requirement for 'orphaned' tokens, which tracks tokens until they are revoked by the `logout()` command or expired.

**Comment by NickolausDS** [comment]
@jimPruyne @ericblau As far as I can tell, this is an overview of the problems we wanted to discuss later today. I very likely missed stuff, since both of you have a much better understanding of the requirements for automate.

**Comment by NickolausDS** [comment]
We previously discussed storing scope sets as a third alternative to the two approaches above. Scope sets would track all of the scopes and tokens for a given Globus auth flow. If another login happens, both entire token sets would be stored as another 'scope set'. The advantage here is no tokens will be 'lost' by new tokens overwriting them, and older tokens (which may be in use) won't need to be revoked. Finally, all tokens are tracked and so all tokens can be revoked by a logout. 

I'm defining "Scope Set" as a set of Globus Scopes

    {'openid', 'profile', 'urn:globus:auth:scope:transfer.api.globus.org:all'}

Some additional considerations:

* `load_tokens` would be ambiguous. The user may have done several login flows with different scope sets.
    * Alternatively, `load_tokens()` with no arguments could load tokens from the latest auth flow.
    * `load_tokens(requested_scopes=)` could load a specific scope set if it exists.
* `get_authorizers` suffers the same ambiguity, but could follow the same logic as `load_tokens`

This change also solves the problem of requesting different scopes for a resource server in-between login flows. It isn't possible to have multiple tokens for the same resource server for a scope set, so listing tokens both by 'resource server' and by 'scopes' will continue to work normally. 

#24 @jimPruyne, @ericblau -- This follows our discussion Friday. Looking over the details, I'm pretty sure this fixes all the token handling problems we discussed. Let me know if anything above diverged from your thinking.

**Comment by NickolausDS** [comment]
Some of the concerns were addressed here in #33. Additionally since then, #41 added fetching tokens and authorizers by scope instead of by resource server. At this point, the underlying problem has been fixed and we can continue storing tokens by resource server.
